Assets/Scripts/Enemy.cs

```diff
@@ -2,34 +2,57 @@
+    protected ENEMY_STATE _state = ENEMY_STATE.idle;
+    // By initializing this as true, we ensure an enemy can change state on its first frame
+    protected bool _can_change_state = true;
 
-    private Rigidbody2D myRigidbody;
-    private SpriteRenderer mySprite;
-    public Rigidbody2D player;
-    // Start is called before the first frame update
-    public float max_speed = 4;
-    public float acceleration_rate = 1;
-    private void Awake(){
-        myRigidbody = GetComponent<Rigidbody2D>();
-        mySprite = GetComponent<SpriteRenderer>();
-    }
-    void Start()
-    {
-        player = GameObject.FindGameObjectWithTag("Player").GetComponent<Rigidbody2D>();
+    // We assume all enemies will have a Rigidbody2D for collision purposes
+    protected Rigidbody2D _myRigidbody;
+
+    // We enforce that an enemy define a Start method to initialize its enemy-specific behaviours
+    protected abstract void Start();
+
+    // We perform updates to enemies on FixedUpdate since they are physics objects
+    // FixedUpdate is virtual to allow overriding by custom non-state-based enemies (such as bosses, potentially)
+    protected virtual void FixedUpdate() {
+        if(this._can_change_state) {
+            this.ChangeState();
+        }
+        try {
+            this._can_change_state = this.PerformAction();
+        } catch(InvalidEnemyStateException ex) {
+            Debug.LogException(ex);
+            this._state = ENEMY_STATE.idle;
+            this._can_change_state = true;
+        }
     }
 
-    // Update is called once per frame
-    void Update()
-    {
-        // Used to normalize updates across various frame-rates
-        float time_step = Time.fixedDeltaTime;
+    // We currently enforce that an enemy define a ChangeState method to determine the enemy-specific conditions under which it becomes idle, aware, and engaged.
+    protected abstract void ChangeState();
 
-        // Constructs a unit-vector along one of the eight digital directions
-        Vector2 acceleration = player.position - myRigidbody.position;
-        mySprite.flipX = acceleration.x <= 0;
-        acceleration *= time_step * acceleration_rate / acceleration.magnitude;
-        myRigidbody.velocity = Vector2.ClampMagnitude(myRigidbody.velocity + acceleration, max_speed);
+    // PerformAction is virtual to allow overriding by custom non-state-based enemies (such as bosses, potentially)
+    protected virtual bool PerformAction() {
+        switch(this._state) {
+            case ENEMY_STATE.idle:
+                return this._idleBehaviour.Act();
+            case ENEMY_STATE.aware:
+                return this._awareBehaviour.Act();
+            case ENEMY_STATE.engaged:
+                return this._engagedBehaviour.Act();
+            default:
+                throw new InvalidEnemyStateException("INVALID ENEMY STATE ERROR: " + this.GetType() + " has a non-standard enemy state while using base PerformAction function.");
+        }
```

Comment:
imo we should prefer using `Dictonary` types over switches and complex conditionals whenever possible. Any way to avoid branching is preferred for readability and performance.

Sometimes branching is necessary, when the logical operations required are genuinely different and cannot be factored out, but here I think we can safely and effectively replace branching.

Reply:
Replaced switch with dictionary lookup. Suggestion accepted.